File: 2015/java/8/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Main {
    // countChars calculates the number of characters in the code and the
    // number of characters in the string representation, excluding escape
    // sequences.
    public static void countChars(String s, Counter codeChars, Counter stringChars) {
        codeChars.value += s.length();

        for (int i = 1; i < s.length() - 1; i++) { // Skip first and last quote
            if (s.charAt(i) == '\\') {
                if (i + 1 < s.length() - 1) {
                    switch (s.charAt(i + 1)) {
                        case '\\':
                        case '"':
                            stringChars.value++;
                            i++;
                            break;
                        case 'x':
                            if (i + 3 < s.length() - 1) {
                                stringChars.value++;
                                i += 3;
                            }
                            break;
                        default:
                            stringChars.value++;
                    }
                }
            } else {
                stringChars.value++;
            }
        }
    }

    // encodeString returns an encoded version of the input string and its length.
    public static EncodedString encodeString(String s) {
        StringBuilder b = new StringBuilder();
        b.append("\"");

        for (char ch : s.toCharArray()) {
            if (ch == '"' || ch == '\\') {
                b.append('\\');
            }
            b.append(ch);
        }

        b.append("\"");
        return new EncodedString(b.toString(), b.length());
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("missing input file");
            System.exit(1);
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(args[0]))) {
            int totalCodeChars = 0;
            int totalStringChars = 0;
            int totalOrigChars = 0;
            int totalEncChars = 0;

            String line;
            while ((line = reader.readLine()) != null) {
                // Part 1: Count characters
                Counter codeChars = new Counter();
                Counter stringChars = new Counter();
                countChars(line, codeChars, stringChars);
                totalCodeChars += codeChars.value;
                totalStringChars += stringChars.value;

                // Part 2: Encode string
                totalOrigChars += line.length();
                EncodedString encString = encodeString(line);
                totalEncChars += encString.length;
            }

            System.out.printf("Part 1: %d\n", totalCodeChars - totalStringChars);
            System.out.printf("Part 2: %d\n", totalEncChars - totalOrigChars);
        } catch (IOException e) {
            System.err.printf("cannot open file: %v\n", e);
            System.exit(1);
        }
    }
}

class Counter {
    int value = 0;
}

class EncodedString {
    String encoded;
    int length;

    EncodedString(String encoded, int length) {
        this.encoded = encoded;
        this.length = length;
    }
}

File: 2015/rust/8/src/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};

// Count the characters in the code and the number of characters in the string representation.
fn count_chars(s: &str, code_chars: &mut usize, string_chars: &mut usize) {
    *code_chars += s.len();

    let bytes = s.as_bytes();
    let mut i = 1; // Skip first quote

    while i < bytes.len() - 1 {
        // Skip last quote
        if bytes[i] == b'\\' {
            if i + 1 < bytes.len() - 1 {
                match bytes[i + 1] {
                    b'\\' | b'"' => {
                        *string_chars += 1;
                        i += 1;
                    }
                    b'x' => {
                        if i + 3 < bytes.len() - 1 {
                            *string_chars += 1;
                            i += 3;
                        }
                    }
                    _ => *string_chars += 1,
                }
            }
        } else {
            *string_chars += 1;
        }
        i += 1;
    }
}

// Encode the string with escaping and return the encoded string and its length.
fn encode_string(s: &str) -> (String, usize) {
    let mut encoded = String::new();
    encoded.push('"');

    for ch in s.chars() {
        if ch == '"' || ch == '\\' {
            encoded.push('\\');
        }
        encoded.push(ch);
    }

    encoded.push('"');
    (encoded.clone(), encoded.len()) // Return a clone of the encoded string and its length
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("missing input file");
        std::process::exit(1);
    }

    let path = &args[1];
    let file = File::open(path)?;
    let reader = io::BufReader::new(file);

    let mut total_code_chars = 0;
    let mut total_string_chars = 0;
    let mut total_orig_chars = 0;
    let mut total_enc_chars = 0;

    for line in reader.lines() {
        let line = line?;
        let line = line.trim();

        // Part 1: Count characters
        count_chars(line, &mut total_code_chars, &mut total_string_chars);

        // Part 2: Encode string
        total_orig_chars += line.len();

        let (_, enc_len) = encode_string(line);
        total_enc_chars += enc_len;
    }

    println!("Part 1: {}", total_code_chars - total_string_chars);
    println!("Part 2: {}", total_enc_chars - total_orig_chars);

    Ok(())
}
